### ISSUE TITLE
Replace incorreto

### ISSUE THREAD

**Issue by ihavenonickname** [enredo/enredo#14]
O seguinte script

    variavel x <- 'é melhor jair se acostumando'
    exibir('1 igual a 1')

Gera a seguinte saída

    const prompt = require('prompt-sync')()

    var x = 'é melhor jair if acostumando'
    console.log('1 == a 1')

Ou seja, é preciso que o programa faça um `replace` mais sofisticado, que substitua os tokens somente quando necessário.

**Comment by eptaccio** [comment]
Entendi o problema, obrigado por reportar.

**Comment by ihavenonickname** [comment]
A minha sugestão é criar um [LL parser](https://en.wikipedia.org/wiki/LL_parser) em vez de usar um mero replace, senão você sempre vai se deparar com algum tipo de limitação como a que reportei.

**Comment by eptaccio** [comment]
@ihavenonickname já implementou isso alguma vez?

**Comment by ihavenonickname** [comment]
Sim, [neste projeto](https://github.com/ihavenonickname/odyssey) e [neste outro projeto](https://github.com/ihavenonickname/Mythology) também.

**Comment by eptaccio** [comment]
massa, consegue dar uma força? parece bem complexo

**Comment by ihavenonickname** [comment]
O código é bem straightforward uma vez que a teoria é entendida.

Um parser se divide em duas fases:

* **Análise léxica**: Transformar um stream de caracteres (texto) em um stream de tokens. No caso do meu projeto Mythology, o stream de tokens é uma lista de objetos. Um token é uma unidade léxica de uma dada linguagem, e sempre está associada a um lexema, que é a aparição de um token no texto.
* **Análise sintática**: Verificar se o stream de tokens corresponde a um programa válido na sua gramática. A maneira mais simples de fazer isso é criar uma rotina para cada regra da gramática da linguagem, chamando recursivamente outras rotinas quando apropriado. É bem simples de fazer, mas a gramática precisa estar devidamente preparada para ser codificada. Enquanto a análise sintática é feita, deve-se montar uma AST do código.

Uma vez que a análise sintática for finalizada a AST estará completa. Basta visitar todos os nodos e ir gerando o texto correspondente a cada nodo (essa é a parte que gera o código JavaScript).

Como você se propôs a criar uma linguagem de programação, imagino que já tenha estudado compiladores. Se ainda não estiver bem afiado nessa matéria, recomendo os PDFs [desta página](https://web.stanford.edu/class/archive/cs/cs143/cs143.1128/). São bem legais e me ajudaram a aprender.

Se precisar de mais ajuda, pode contar comigo, mas eu não sou nenhum especialista na área.

**Comment by pglapds** [comment]
> 
> 
> O código é bem straightforward uma vez que a teoria é entendida.
> 
> Um parser se divide em duas fases:
> 
>     * **Análise léxica**: Transformar um stream de caracteres (texto) em um stream de tokens. No caso do meu projeto Mythology, o stream de tokens é uma lista de objetos. Um token é uma unidade léxica de uma dada linguagem, e sempre está associada a um lexema, que é a aparição de um token no texto.
> 
>     * **Análise sintática**: Verificar se o stream de tokens corresponde a um programa válido na sua gramática. A maneira mais simples de fazer isso é criar uma rotina para cada regra da gramática da linguagem, chamando recursivamente outras rotinas quando apropriado. É bem simples de fazer, mas a gramática precisa estar devidamente preparada para ser codificada. Enquanto a análise sintática é feita, deve-se montar uma AST do código.
> 
> 
> Uma vez que a análise sintática for finalizada a AST estará completa. Basta visitar todos os nodos e ir gerando o texto correspondente a cada nodo (essa é a parte que gera o código JavaScript).
> 
> Como você se propôs a criar uma linguagem de programação, imagino que já tenha estudado compiladores. Se ainda não estiver bem afiado nessa matéria, recomendo os PDFs [desta página](https://web.stanford.edu/class/archive/cs/cs143/cs143.1128/). São bem legais e me ajudaram a aprender.
> 
> Se precisar de mais ajuda, pode contar comigo, mas eu não sou nenhum especialista na área.

Interessante.

**Comment by phiberber** [comment]
Quando olhei o tanto de arquivos no projeto já percebi que era baseado em Regex. Se você não quiser obrigatoriamente fazer um Parser, um Lexer e Tokenizer simples já resolveriam este problema, você poderia substituir apenas um Token.SE por um "if".
